File: sim/aether_des.js
/**
 * AETHER Web-Lite: 離散事象シミュレータ (Discrete Event Simulator)
 * 
 * 暗号処理などの重い処理をバイパスし、純粋なネットワークトポロジ、
 * ルーティング、遅延、ノード離脱（Churn）などの「物理現象」をシミュレートします。
 */

// ============================================================
// 1. シミュレーター設定パラメータ
// ============================================================
const CONFIG = {
    NUM_NODES: process.argv[2] ? parseInt(process.argv[2], 10) : 2000, // 総ノード数 (引数で指定可能)
    NUM_ZONES: 256,           // 総Zone数
    ZONES_PER_NODE: 16,       // 1ノードが参加するZone数
    DEGREE: 6,                // Gossipメッシュの接続数（fanout）
    K_REPLICATION: 5,         // Mailboxへの保管ノード数
    BASE_LATENCY_MS: 50,      // 基準通信遅延（Ping）
    JITTER_MS: 50,            // 遅延の揺らぎ（0〜50ms）
    CHURN_RATE: 0.20,         // ネットワーク切断率（途中で20%が突然死する）
};

// ============================================================
// 2. 離散事象エンジン (Priority Queue)
// ============================================================
class PriorityQueue {
    constructor() { this.events = []; }
    push(e) { 
        this.events.push(e); 
        this.events.sort((a,b) => a.time - b.time); 
    }
    pop() { return this.events.shift(); }
    isEmpty() { return this.events.length === 0; }
}

const sim = {
    time: 0,
    q: new PriorityQueue(),
    schedule: (delay, action) => {
        sim.q.push({ time: sim.time + delay, action });
    },
    run: () => {
        let eventCount = 0;
        while(!sim.q.isEmpty()) {
            const e = sim.q.pop();
            sim.time = e.time;
            e.action();
            eventCount++;
        }
        return eventCount;
    }
};

// ============================================================
// 3. ネットワークとノード定義
// ============================================================
class Node {
    constructor(id) {
        this.id = id;
        this.isOnline = true;
        this.zones = new Set();               // 所属する16個のZone
        this.mesh = new Map();                // ZoneID -> Set<NodeID> (隣接ピア)
        this.mailbox = new Set();             // 保管しているパケットID (DHT)
        this.seenPackets = new Set();         // ゴシップでの重複排除用
    }

    // ネットワーク遅延をシミュレートしてパケットを送信
    send(targetNode, packet, handler) {
        if (!this.isOnline || !targetNode.isOnline) return;
        const latency = CONFIG.BASE_LATENCY_MS + Math.random() * CONFIG.JITTER_MS;
        sim.schedule(latency, () => {
            if (targetNode.isOnline) handler(targetNode, packet);
        });
    }
}

const network = {
    nodes: [],
    zoneMap: new Map(), // ZoneID -> Set<NodeID>
    metrics: {
        totalGossipSent: 0,
        totalGossipReceived: 0,
        mailboxHits: 0,
        deliveries: new Map(), // packetId -> [{nodeId, time}]
    }
};

// ランダム選択ユーティリティ (O(K) に最適化)
const getRandomElements = (arr, count) => {
    if (count >= arr.length) return arr;
    const result = new Set();
    while(result.size < count) {
        result.add(arr[Math.floor(Math.random() * arr.length)]);
    }
    return Array.from(result);
};

const getLatency = () => CONFIG.BASE_LATENCY_MS + Math.random() * CONFIG.JITTER_MS;

// ============================================================
// 4. メインルーチン
// ============================================================

console.log("╔═══════════════════════════════════════════════════════════════════╗");
console.log("║     AETHER Web-Lite: ネットワーク動的シミュレーター              ║");
console.log("╚═══════════════════════════════════════════════════════════════════╝");
console.log(`設定: ${CONFIG.NUM_NODES} nodes, ${CONFIG.NUM_ZONES} zones, Churn: ${CONFIG.CHURN_RATE*100}%, Delay: ${CONFIG.BASE_LATENCY_MS}ms\n`);

// [Step 1] ノードの初期化とZoneの割り当て
console.log("[Phase 1] ネットワークの自律配備 (Bootstrap)...");
for (let i = 0; i < CONFIG.NUM_NODES; i++) {
    const node = new Node(i);
    // 256のうち16個のZoneをランダムに選択
    const allZones = Array.from({length: CONFIG.NUM_ZONES}, (_, k) => k);
    const selectedZones = getRandomElements(allZones, CONFIG.ZONES_PER_NODE);
    
    selectedZones.forEach(z => {
        node.zones.add(z);
        if (!network.zoneMap.has(z)) network.zoneMap.set(z, new Set());
        network.zoneMap.get(z).add(i);
    });
    network.nodes.push(node);
}

// 参照速度の最適化のため配列キャッシュを作成
network.zoneArrMap = new Map();
network.zoneMap.forEach((set, z) => network.zoneArrMap.set(z, Array.from(set)));

// [Step 2] libp2p-gossipsub メッシュの自律形成
for (let i = 0; i < CONFIG.NUM_NODES; i++) {
    const node = network.nodes[i];
    node.zones.forEach(z => {
        node.mesh.set(z, new Set());
        // IDが自分と同じになるのを防ぐための簡易フィルタ
        let occupants = network.zoneArrMap.get(z);
        if (occupants.length <= 1) return;
        
        let attempts = 0;
        while(node.mesh.get(z).size < Math.min(CONFIG.DEGREE, occupants.length - 1) && attempts < 20) {
            const p = occupants[Math.floor(Math.random() * occupants.length)];
            if (p !== node.id) node.mesh.get(z).add(p);
            attempts++;
        }
        const peers = Array.from(node.mesh.get(z));
        // 双方向リンクの構築（WebRTC）
        peers.forEach(p => {
            if (!network.nodes[p].mesh.has(z)) network.nodes[p].mesh.set(z, new Set());
            network.nodes[p].mesh.get(z).add(node.id);
        });
    });
}

// 分析: メッシュの形成状態
let emptyZones = 0;
let avgOccupants = 0;
for (let z = 0; z < CONFIG.NUM_ZONES; z++) {
    const size = network.zoneMap.get(z)?.size || 0;
    avgOccupants += size;
    if (size === 0) emptyZones++;
}
console.log(`  - 完了: 1Zoneあたりの平均住民数 = ${(avgOccupants/CONFIG.NUM_ZONES).toFixed(1)}人`);
console.log(`  - 空のZone数: ${emptyZones} / ${CONFIG.NUM_ZONES}\n`);


// [Step 3] パケット配信ロジック定義
const processStem = (node, packet) => {
    // 10%でFluff（拡散）へ移行、90%でStem継続
    if (Math.random() < 0.10) {
        // Fluff: ターゲットのZoneにInjectする
        processFluffInject(node, packet);
    } else {
        // Stem: ランダムな隣接ピア（全接続中から1つ）へバケツリレー
        const allPeers = new Set();
        node.mesh.forEach(peers => peers.forEach(p => {
            if (network.nodes[p].isOnline) allPeers.add(p);
        }));
        if (allPeers.size === 0) return processFluffInject(node, packet); // 逃げ道
        
        const randomPeerId = getRandomElements(Array.from(allPeers), 1)[0];
        const randomPeer = network.nodes[randomPeerId];
        node.send(randomPeer, packet, processStem);
    }
};

const processFluffInject = (flufferNode, packet) => {
    // 1. DHT Mailboxへの保存 (非同期)
    // モック: パケットIDに最も近いKノードをMailboxとする
    const kClosest = network.nodes.slice().sort((a,b) => 
        Math.abs(a.id - packet.targetHash) - Math.abs(b.id - packet.targetHash)
    ).slice(0, CONFIG.K_REPLICATION);

    kClosest.forEach(target => {
        flufferNode.send(target, packet, (n, p) => {
            n.mailbox.add(p.id);
            network.metrics.mailboxHits++;
        });
    });

    // 2. 対象Zoneのメッシュへの投入
    // Flufferが対象Zoneに繋がっていなければ、DHTで誰かを探して投下する
    let injected = false;
    if (flufferNode.zones.has(packet.zoneId)) {
        processGossip(flufferNode, packet);
        injected = true;
    } else {
        const zoneMembers = network.zoneArrMap.get(packet.zoneId) || [];
        // Extract random online member
        let targetMember = null;
        for (let i = 0; i < 20; i++) {
            const candidateId = zoneMembers[Math.floor(Math.random() * zoneMembers.length)];
            if (network.nodes[candidateId] && network.nodes[candidateId].isOnline) {
                targetMember = network.nodes[candidateId];
                break;
            }
        }
        
        if (targetMember) {
            flufferNode.send(targetMember, packet, processGossip);
            injected = true;
        }
    }
};

const processGossip = (node, packet) => {
    if (packet.targetZone !== undefined && packet.targetZone !== packet.zoneId) return; // fail-safe
    if (node.seenPackets.has(packet.id)) return; // 重複排除
    
    node.seenPackets.add(packet.id);
    network.metrics.totalGossipReceived++;

    // 計測: 受け取った時間を記録 (作者以外)
    if (node.id !== packet.authorId) {
        if (!network.metrics.deliveries.has(packet.id)) {
            network.metrics.deliveries.set(packet.id, []);
        }
        network.metrics.deliveries.get(packet.id).push({ nodeId: node.id, time: sim.time });
    }

    // Gossip中継 (fanout)
    const peers = Array.from(node.mesh.get(packet.zoneId) || []);
    // fanout=4の確率的ゴシップ
    const targets = getRandomElements(peers, 4);
    
    targets.forEach(peerId => {
        network.metrics.totalGossipSent++;
        node.send(network.nodes[peerId], packet, processGossip);
    });
};

// テスト実行関数
const runPublishTest = (testName, targetZone) => {
    // Reset specific metrics for this run
    network.metrics.totalGossipSent = 0;

    console.log(`[Phase] ${testName} を開始...`);
    const packetId = Math.floor(Math.random() * 1000000);
    const onlineNodes = network.nodes.filter(n => n.isOnline);
    const author = onlineNodes[Math.floor(Math.random() * onlineNodes.length)];
    
    // 作者(author)はTarget Zoneにいなくても書き込める (Dandelion++)
    const packet = {
        id: packetId,
        zoneId: targetZone,
        authorId: author.id,
        targetHash: Math.floor(Math.random() * CONFIG.NUM_NODES), // for Mailbox routing
        postedAt: sim.time
    };

    // 発射！ (作者がStemフェーズを開始)
    sim.schedule(0, () => processStem(author, packet));
    sim.run();

    // 結果の集計
    const targetMembers = Array.from(network.zoneMap.get(targetZone) || []).filter(id => network.nodes[id].isOnline);
    const expectedCount = targetMembers.length - (author.zones.has(targetZone) ? 1 : 0);
    
    const deliveries = network.metrics.deliveries.get(packetId) || [];
    const deliveredCount = deliveries.length;
    const deliveryRate = expectedCount > 0 ? ((deliveredCount / expectedCount) * 100).toFixed(1) : 100;

    let maxDelay = 0;
    let sumDelay = 0;
    deliveries.forEach(d => {
        const delay = d.time - packet.postedAt;
        if (delay > maxDelay) maxDelay = delay;
        sumDelay += delay;
    });
    const avgDelay = deliveredCount > 0 ? (sumDelay / deliveredCount).toFixed(0) : 0;

    console.log(`  - ターゲットの生還ノード数: ${expectedCount}人`);
    console.log(`  - 配達成功数: ${deliveredCount}人 (${deliveryRate}%)`);
    if (deliveredCount > 0) {
        console.log(`  - 遅延 (作者から全ノードへ): 平均 ${avgDelay} ms, 最大 ${maxDelay.toFixed(0)} ms`);
    } else {
        console.log(`  - 誰もいない！ (DHTにのみ格納されました)`);
    }
    console.log(`  - 消費帯域 (1件のメッセージで発生した全通信回数): Gossip ${network.metrics.totalGossipSent} 回`);
    console.log("");
};


// [Step 4] シナリオ実行

// シナリオ1: 安定状態での送信
runPublishTest("通常テスト(安定状態・人気板 Zone12 への書き込み)", 12);

// シナリオ2: 過疎板への送信
runPublishTest("過疎板テスト(Zone230 への書き込み)", 230);

// [Step 5] ネットワーク切断テロ (Churn) の発生
console.log(`[Phase] ⚠️ 突然のネットワーク障害発生！ (Churn Rate: ${CONFIG.CHURN_RATE*100}%)`);
let offlineCount = 0;
network.nodes.forEach(node => {
    if (Math.random() < CONFIG.CHURN_RATE) {
        node.isOnline = false;
        offlineCount++;
    }
});
console.log(`  - ${offlineCount} 個のノードが突然ダウン（切断・タブ閉じ）しました。メッシュの修復は行いません。\n`);

// シナリオ3: 障害発生直後の送信 (自己修復を待たずに送信)
// Meshは物理的に寸断されている可能性があるが、冗長性(degree=6/fanout=4)でどこまで耐えるか？
runPublishTest("障害直後テスト(切断状態で人気板 Zone12 への書き込み)", 12);

console.log("╔═══════════════════════════════════════════════════════════════════╗");
console.log("║                       シ ミ ュ レ ー シ ョ ン 完 了               ║");
console.log("╚═══════════════════════════════════════════════════════════════════╝");

File: sim/aether_des.cpp
#include <iostream>
#include <vector>
#include <random>
#include <queue>
#include <algorithm>
#include <chrono>

using namespace std;

const int NUM_ZONES = 256;
const int ZONES_PER_NODE = 16;
const int DEGREE = 6;
const int FANOUT = 4;
const int BASE_LATENCY_MS = 50;
const int JITTER_MS = 50;

struct Event {
    int time;
    int type; // 0=STEM, 1=FLUFF, 2=GOSSIP
    int node_id;
    int zone_id;
    int packet_id;
    int author_id;
    bool operator>(const Event& other) const { return time > other.time; }
};

struct Node {
    bool is_online = true;
    uint8_t zones[ZONES_PER_NODE];
    uint8_t mesh_counts[ZONES_PER_NODE] = {0};
    int mesh[ZONES_PER_NODE][10]; 
    int seen_packet_id = -1; 
};

struct Metrics {
    int deliveries = 0;
    long long sum_latency = 0;
    int max_latency = 0;
    long long total_gossip = 0;
};

int main(int argc, char** argv) {
    int NUM_NODES = 1000000;
    if (argc > 1) NUM_NODES = atoi(argv[1]);
    
    double CHURN_RATE = 0.2;
    if (argc > 2) CHURN_RATE = atof(argv[2]);

    cout << "\n========================================================" << endl;
    cout << "   AETHER Web-Lite C++ Global P2P Simulator" << endl;
    cout << "========================================================" << endl;
    cout << "Nodes: " << NUM_NODES << " | Zones: 256 | Degree: 6 | Fanout: 4 | Churn: " << CHURN_RATE*100 << "%" << endl;
    
    auto t1 = chrono::steady_clock::now();
    
    vector<Node>* nodes = new vector<Node>(NUM_NODES);
    vector<vector<int>> zoneMap(NUM_ZONES);
    
    mt19937 gen(42);
    
    cout << "\n[Phase] Initializing Zones (Bootstrap)... ";
    vector<uint8_t> all_z(NUM_ZONES);
    for(int z=0; z<NUM_ZONES; z++) all_z[z] = z;
    
    for (int i = 0; i < NUM_NODES; ++i) {
        shuffle(all_z.begin(), all_z.end(), gen);
        for (int j = 0; j < ZONES_PER_NODE; ++j) {
            (*nodes)[i].zones[j] = all_z[j];
            zoneMap[all_z[j]].push_back(i);
        }
    }
    cout << "Done." << endl;
    
    cout << "[Phase] Building Global Mesh Graph... ";
    for (int z = 0; z < NUM_ZONES; ++z) {
        auto& occupants = zoneMap[z];
        for (int node_id : occupants) {
            int z_idx = -1;
            for(int i=0; i<ZONES_PER_NODE; i++) {
                if ((*nodes)[node_id].zones[i] == z) { z_idx = i; break; }
            }
            if (z_idx == -1) continue;
            
            int attempts = 0;
            while ((*nodes)[node_id].mesh_counts[z_idx] < DEGREE && attempts < 20) {
                attempts++;
                int p = occupants[gen() % occupants.size()];
                if (p == node_id) continue;
                
                bool exists = false;
                for(int k=0; k<(*nodes)[node_id].mesh_counts[z_idx]; k++) {
                    if((*nodes)[node_id].mesh[z_idx][k] == p) exists = true;
                }
                if (exists) continue;

                (*nodes)[node_id].mesh[z_idx][(*nodes)[node_id].mesh_counts[z_idx]++] = p;
                
                int p_z_idx = -1;
                for(int i=0; i<ZONES_PER_NODE; i++) {
                    if ((*nodes)[p].zones[i] == z) { p_z_idx = i; break; }
                }
                if (p_z_idx != -1 && (*nodes)[p].mesh_counts[p_z_idx] < 10) {
                    bool p_exists = false;
                    for(int k=0; k<(*nodes)[p].mesh_counts[p_z_idx]; k++) {
                        if((*nodes)[p].mesh[p_z_idx][k] == node_id) p_exists = true;
                    }
                    if (!p_exists) {
                        (*nodes)[p].mesh[p_z_idx][(*nodes)[p].mesh_counts[p_z_idx]++] = node_id;
                    }
                }
            }
        }
    }
    
    auto t2 = chrono::steady_clock::now();
    cout << "Done. (" << chrono::duration_cast<chrono::milliseconds>(t2 - t1).count() << " ms)" << endl;
    
    auto runTest = [&](string name, int target_z, int start_time, int packet_id) {
        cout << "\n[Test] " << name << " (Zone " << target_z << ")" << endl;
        priority_queue<Event, vector<Event>, greater<Event>> pq;
        Metrics metrics;
        
        int author = -1;
        while(true) {
            author = gen() % NUM_NODES;
            if ((*nodes)[author].is_online) break;
        }
        
        pq.push({start_time, 0, author, target_z, packet_id, author});
        
        while(!pq.empty()) {
            Event ev = pq.top(); pq.pop();
            if (!(*nodes)[ev.node_id].is_online) continue;
            
            if (ev.type == 0) { // STEM
                if ((gen() % 100) < 10) { // Fluff
                    pq.push({ev.time, 1, ev.node_id, ev.zone_id, ev.packet_id, ev.author_id}); 
                } else { // Continue Stem
                    vector<int> available;
                    for(int z_idx=0; z_idx<ZONES_PER_NODE; z_idx++) {
                        for(int i=0; i<(*nodes)[ev.node_id].mesh_counts[z_idx]; i++) {
                            int n = (*nodes)[ev.node_id].mesh[z_idx][i];
                            if ((*nodes)[n].is_online) available.push_back(n);
                        }
                    }
                    if (available.empty()) {
                        pq.push({ev.time, 1, ev.node_id, ev.zone_id, ev.packet_id, ev.author_id}); 
                    } else {
                        int next_node = available[gen() % available.size()];
                        int delay = BASE_LATENCY_MS + (gen() % JITTER_MS);
                        pq.push({ev.time + delay, 0, next_node, ev.zone_id, ev.packet_id, ev.author_id});
                    }
                }
            } else if (ev.type == 1) { // FLUFF
                int start_node = ev.node_id;
                bool in_zone = false;
                for(int i=0; i<ZONES_PER_NODE; i++) {
                    if ((*nodes)[ev.node_id].zones[i] == ev.zone_id) in_zone = true;
                }
                if (!in_zone) {
                    const auto& mems = zoneMap[ev.zone_id];
                    for(int i=0; i<50; i++) {
                        int candidate = mems[gen() % mems.size()];
                        if ((*nodes)[candidate].is_online) {
                            start_node = candidate;
                            break;
                        }
                    }
                }
                pq.push({ev.time, 2, start_node, ev.zone_id, ev.packet_id, ev.author_id});
            } else if (ev.type == 2) { // GOSSIP
                if ((*nodes)[ev.node_id].seen_packet_id == ev.packet_id) continue;
                (*nodes)[ev.node_id].seen_packet_id = ev.packet_id;
                
                if (ev.node_id != ev.author_id) {
                    metrics.deliveries++;
                    metrics.max_latency = max(metrics.max_latency, ev.time - start_time);
                    metrics.sum_latency += (ev.time - start_time);
                }
                
                int z_idx = -1;
                for(int i=0; i<ZONES_PER_NODE; i++) {
                    if ((*nodes)[ev.node_id].zones[i] == ev.zone_id) { z_idx = i; break; }
                }
                if (z_idx != -1) {
                    vector<int> neighbors;
                    for(int i=0; i<(*nodes)[ev.node_id].mesh_counts[z_idx]; i++) {
                        int n = (*nodes)[ev.node_id].mesh[z_idx][i];
                        if ((*nodes)[n].is_online) neighbors.push_back(n);
                    }
                    shuffle(neighbors.begin(), neighbors.end(), gen);
                    int count = 0;
                    for(int n : neighbors) {
                        int delay = BASE_LATENCY_MS + (gen() % JITTER_MS);
                        pq.push({ev.time + delay, 2, n, ev.zone_id, ev.packet_id, ev.author_id});
                        metrics.total_gossip++;
                        if (++count >= FANOUT) break;
                    }
                }
            }
        }
        
        int expected = 0;
        for (int n : zoneMap[target_z]) {
            if ((*nodes)[n].is_online && n != author) expected++;
        }
        
        double rate = expected > 0 ? (double)metrics.deliveries / expected * 100.0 : 100.0;
        int avg_lat = metrics.deliveries > 0 ? (metrics.sum_latency / metrics.deliveries) : 0;
        
        // 帯域計算 (1秒間あたりの推定全ユーザー平均使用量)
        long long global_posts_per_sec = NUM_NODES / 200; // 例: 100万人なら秒間5000件
        if (global_posts_per_sec == 0) global_posts_per_sec = 1;
        
        double avg_msgs_per_post = (double)metrics.total_gossip / NUM_NODES;
        double avg_kbps = (avg_msgs_per_post * global_posts_per_sec * 500.0) / 1024.0;

        cout << "  * Expected Target Nodes: " << expected << endl;
        cout << "  * Delivered Copies: " << metrics.deliveries << " (" << rate << "%)" << endl;
        if (metrics.deliveries > 0) {
            cout << "  * Latency: Avg " << avg_lat << " ms, Max " << metrics.max_latency << " ms" << endl;
        } else {
            cout << "  * DHT MAILBOX Storage Only (No real-time delivery)" << endl;
        }
        cout << "  * Total Gossip Msgs: " << metrics.total_gossip << " (for 1 post)" << endl;
        cout << "  * Global Mails/Sec: " << global_posts_per_sec << " posts/sec" << endl;
        cout << "  * Avg Bandwidth/Node: " << avg_kbps << " KB/s  (推定・全ユーザー平均)" << endl;
    };
    
    runTest("Normal State Stability", 42, 0, 100);
    
    cout << "\n[Phase] Triggering " << CHURN_RATE*100 << "% Network Churn Attack..." << endl;
    int offline_count = 0;
    for(int i=0; i<NUM_NODES; i++) {
        if ((gen() % 100) < (CHURN_RATE * 100)) {
            (*nodes)[i].is_online = false;
            offline_count++;
        }
    }
    cout << "  " << offline_count << " nodes went instantly offline." << endl;
    
    runTest("Post-Churn Resilience", 42, 10000, 101);
    
    delete nodes;
    return 0;
}
